File: rust/alera-cli/src/issue_tracking/issue_details.rs
use serde::{Deserialize, Serialize};

use super::IssueProvider;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum IssueState {
    Open,
    Closed,
    Unknown,
}

impl IssueState {
    pub fn wire_name(self) -> &'static str {
        match self {
            IssueState::Open => "open",
            IssueState::Closed => "closed",
            IssueState::Unknown => "unknown",
        }
    }
}

/// The provider-neutral issue payload returned by `issue.fetch` and printed by
/// `alera issue show --json`. `state_label` keeps the forge's own wording
/// (Azure DevOps states such as `Active` or `Resolved` are process-defined).
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct IssueDetails {
    pub provider: IssueProvider,
    pub url: String,
    #[serde(default)]
    pub repository: Option<String>,
    pub number: i64,
    pub title: String,
    pub state: IssueState,
    #[serde(default)]
    pub state_label: Option<String>,
    #[serde(default)]
    pub body: Option<String>,
    #[serde(default)]
    pub labels: Vec<String>,
    #[serde(default)]
    pub assignees: Vec<String>,
    #[serde(default)]
    pub author: Option<String>,
    #[serde(default)]
    pub created_at: Option<String>,
    #[serde(default)]
    pub updated_at: Option<String>,
}

File: lib/src/features/linked_issues/domain/issue_details.dart
import 'package:alera/src/features/linked_issues/domain/issue_state.dart';
import 'package:alera/src/shared/git_hosting/domain/git_hosting_provider.dart';
import 'package:dart_mappable/dart_mappable.dart';

part 'issue_details.mapper.dart';

/// An issue as the runtime host fetched it from its forge (`issue.fetch`).
@MappableClass()
class const IssueDetails({
  required this.provider,
  required this.url,
  required this.number,
  required this.title,
  required this.state,
  this.repository,
  this.stateLabel,
  this.body,
  this.labels = const <String>[],
  this.assignees = const <String>[],
  this.author,
  this.createdAt,
  this.updatedAt,
}) with IssueDetailsMappable {
  final GitHostingProvider provider;
  final String url;
  final int number;
  final String title;
  final IssueState state;
  final String? repository;
  final String? stateLabel;
  final String? body;
  final List<String> labels;
  final List<String> assignees;
  final String? author;
  final String? createdAt;
  final String? updatedAt;

  String get displayState => stateLabel ?? state.label;

  factory fromJson(Map<String, Object?> json) =>
      IssueDetailsMapper.fromMap(Map<String, dynamic>.from(json));
}
